### PR TITLE
feat: serial number

### PR DESCRIPTION
**Pull request by akshayitzme**
- [x] SerialNo schema
  <div>
    <details>
     <summary>Enable Serial No in InventorySettings</summary>
      
   </details>
  
   <details>
     <summary>Serial No Page, visible only if SerialNo enabled in InventorySettings</summary>
    
  </details>
  </div>

- [x] Enable Serial No transactions 
  <div>
    <details>
     <summary>Serial No in Stock Movement</summary>
    
  </details>

    <details>
     <summary>Serial No in Stock Ledger, visible only if SerialNo enabled in InventorySettings</summary>
    
  </details>

  </div>
- [x] tests
- [x] documentation : [akshayitzme:serial-no#10](https://github.com/frappe/books_docs/pull/10)

### CODE REVIEW

**Comment by akshayitzme** [comment]
@18alantom could you please review this PR ?

**Comment by akshayitzme** [comment]
@rtdany10 could you please review this PR ?


cc: @18alantom

File: models/helpers.ts [hunk 1]
```diff
@@ -248,6 +248,45 @@ export function getInvoiceStatus(doc: RenderData | Doc): InvoiceStatus {
   return 'Saved';
 }
 
+export function getSerialNoStatusColumn(): ColumnConfig {
+  return {
+    label: t`Status`,
+    fieldname: 'status',
+    fieldtype: 'Select',
+    render(doc) {
+      const status = doc.status as SerialNoStatus;
```
Comment (18alantom) [inline models/helpers.ts:257]: Don't use type assertions. Narrow the type by either using predicate functions or conditionals. [ref](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates)

If `doc.status` is not a `string`, let alone a `SerialNoStatus`, downstream code will have invalid states, color will be undefined.

File: models/helpers.ts [hunk 2]
```diff
@@ -248,6 +248,45 @@ export function getInvoiceStatus(doc: RenderData | Doc): InvoiceStatus {
...
+      const color = serialNoStatusColor[status];
+      const label = getSerialNoStatusText(status as string);
```
Comment (18alantom) [inline models/helpers.ts:259]: Unnecessary type assertion, `SerialNoStatus` is already a `string`.

File: schemas/app/inventory/SerialNo.json [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    {
+      "fieldname": "description",
+      "label": "Description",
+      "placeholder": "Serial Number Description",
+      "fieldtype": "Text"
+    },
+    {
+      "fieldname": "status",
+      "label": "Status",
+      "fieldtype": "Data",
```
Comment (18alantom) [inline schemas/app/inventory/SerialNo.json:30]: Why is this a Data field? It should be a Select field, there are a fixed number options Status can be.

File: schemas/app/inventory/SerialNo.json [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      "default": "Inactive"
+    }
+  ],
+  "quickEditFields": ["item", "description", "party"],
```
Comment (18alantom) [inline schemas/app/inventory/SerialNo.json:34]: Why is `party` here?

File: src/utils/sidebarConfig.ts [hunk 5]
```diff
@@ -96,6 +96,13 @@ async function getInventorySidebar(): Promise<SidebarRoot[]> {
           name: 'stock-balance',
           route: '/report/StockBalance',
         },
+        {
```
Comment (18alantom) [inline src/utils/sidebarConfig.ts:99]: This shouldn't be here, sidebar section for inventory is very crammed.

File: schemas/app/inventory/InventorySettings.json [hunk 6]
```diff
@@ -63,6 +63,12 @@
       "fieldtype": "Check",
       "section": "Features"
     },
+    {
+      "fieldname": "enableSerialNo",
+      "label": "Enable Serial No",
```
Comment (18alantom) [inline schemas/app/inventory/InventorySettings.json:68]: There should be a period after an abbreviation

File: schemas/app/Item.json [hunk 7]
```diff
@@ -138,6 +138,13 @@
       "default": false,
       "section": "Inventory"
     },
+    {
+      "fieldname": "hasSerialNo",
+      "label": "Has Serial No",
```
Comment (18alantom) [inline schemas/app/Item.json:143]: There should be a period after an abbreviation

File: models/inventory/StockTransferItem.ts [hunk 8]
```diff
@@ -215,6 +216,7 @@ export class StockTransferItem extends Doc {
 
   override hidden: HiddenMap = {
     batch: () => !this.fyo.singles.InventorySettings?.enableBatches,
+    serialNo: () => !this.fyo.singles.InventorySettings?.enableSerialNo,
```
Comment (18alantom) [inline models/inventory/StockTransferItem.ts:219]: Should be visible only if the selected item has Serial Nos enabled (same with batches).

File: models/inventory/StockMovementItem.ts [hunk 9]
```diff
@@ -236,6 +237,7 @@ export class StockMovementItem extends Doc {
 
   override hidden: HiddenMap = {
     batch: () => !this.fyo.singles.InventorySettings?.enableBatches,
+    serialNo: () => !this.fyo.singles.InventorySettings?.enableSerialNo,
```
Comment (18alantom) [inline models/inventory/StockMovementItem.ts:240]: Should be visible only if the selected item has Serial Nos enabled (same with batches).

File: models/inventory/PurchaseReceipt.ts [hunk 10]
```diff
@@ -1,11 +1,17 @@
 import { ListViewSettings } from 'fyo/model/types';
 import { getTransactionStatusColumn } from 'models/helpers';
+import { updateSerialNoStatus } from './helpers';
 import { PurchaseReceiptItem } from './PurchaseReceiptItem';
 import { StockTransfer } from './StockTransfer';
 
 export class PurchaseReceipt extends StockTransfer {
   items?: PurchaseReceiptItem[];
 
+  async afterSubmit(): Promise<void> {
```
Comment (18alantom) [inline models/inventory/PurchaseReceipt.ts:10]: Why is this here, the same logic is duplicated across all `StockTransfer` subclasses. This should be handled there.

File: models/inventory/PurchaseReceipt.ts [hunk 11]
```diff
@@ -1,11 +1,17 @@
...
+    await super.afterSubmit();
+    await updateSerialNoStatus(this, this.items!, 'Active');
+  }
+  
   static getListViewSettings(): ListViewSettings {
```
Comment (18alantom) [inline models/inventory/PurchaseReceipt.ts:15]: Why isn't there an `afterCancel` to update serial no status like the other sub classes?

File: models/inventory/helpers.ts [hunk 12]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+      await validateSNPurchaseReceipt(
+        doc,
+        name as string,
+        serialNo,
+        status as string
+      );
+    }
+  }
+}
+
```
Comment (18alantom) [inline models/inventory/helpers.ts:152]: This file is filled with arrow functions. Don't use arrow functions unless you're using them as callbacks.

File: models/inventory/helpers.ts [hunk 13]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+};
+
+export const updateSerialNoStatus = async (
+  doc: Doc,
+  items: StockMovementItem[] | ShipmentItem[],
+  newStatus: string
+) => {
+  for (const item of items) {
+    const serialNos = getSerialNumbers(item.serialNo!);
+    if (!serialNos.length) break;
```
Comment (18alantom) [inline models/inventory/helpers.ts:164]: Don't do this, succinctness hurts readability. Use braces
```javascript
if (!serialNos.length) {
  break
}
```

File: models/inventory/helpers.ts [hunk 14]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+  return serialNo ? serialNo.split('\n') : [];
+};
+
+export const updateSerialNoStatus = async (
+  doc: Doc,
+  items: StockMovementItem[] | ShipmentItem[],
+  newStatus: string
+) => {
+  for (const item of items) {
+    const serialNos = getSerialNumbers(item.serialNo!);
```
Comment (18alantom) [inline models/inventory/helpers.ts:163]: Don't use assertions `!`, I know it's there in many places I'm removing them as and when I come across them.

If it `item.serialNo` is not defined you should be handling it instead of asserting that it is defined. This will lead to errors.

File: models/inventory/helpers.ts [hunk 15]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+      await doc.fyo.db.update(ModelNameEnum.SerialNo, {
+        name: serialNo,
+        status: newStatus,
+      });
+    }
+  }
+};
+
+
+const validateSNMaterialReceipt = async (
```
Comment (18alantom) [inline models/inventory/helpers.ts:176]: Bad naming. It should be `validateMaterialReceiptSerialNumber` this reads like a proper phrase.

File: models/inventory/helpers.ts [hunk 16]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+  doc: Doc,
```
Comment (18alantom) [inline models/inventory/helpers.ts:177]: Bad typing. If you're validating for Material Receipts then only the docs which pertain to material receipts should be passed to this function. Don't use a general type where a narrow type can be used.

File: models/inventory/helpers.ts [hunk 17]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+  status: string
+) => {
+  if (name === undefined) {
+    const values = {
+      name: serialNo,
+      item: doc.item,
+      party: doc.parentdoc?.party as string,
+    };
+    (
+      await doc.fyo.doc
```
Comment (18alantom) [inline models/inventory/helpers.ts:189]: Why are you submitting things in a function that says validate? Validations should not have [side effects](https://en.wikipedia.org/wiki/Side_effect_(computer_science))

File: models/inventory/helpers.ts [hunk 18]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+    throw new ValidationError(
+      t`Serial Number ${serialNo} does not belong to the item ${
+        doc.item! as string
+      }`
+    );
+  }
+};
+
+const validateSNShipment = async (doc: Doc, serialNo: string) => {
+  const { status } = await doc.fyo.db.get(
```
Comment (18alantom) [inline models/inventory/helpers.ts:251]: Should be using `fyo.getValue`

Comment (18alantom) [reply models/inventory/helpers.ts:251]: Other than that, same as above bad naming and bad typing.

File: models/inventory/helpers.ts [hunk 19]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+    ModelNameEnum.SerialNo,
+    serialNo,
+    'status'
+  );
+
+  if (status !== 'Active')
```
Comment (18alantom) [inline models/inventory/helpers.ts:257]: Use braces!

File: models/inventory/helpers.ts [hunk 20]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+const validateSNMaterialIssue = async (
+  doc: Doc,
+  name: string,
+  item: string,
+  serialNo: string,
+  status: string
+) => {
+  if (doc.isCancelled) return;
+
+  if (!name)
```
Comment (18alantom) [inline models/inventory/helpers.ts:234]: Braces!

File: models/inventory/helpers.ts [hunk 21]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+    throw new ValidationError(t`Serial Number ${serialNo} does not exist.`);
+
+  if (status !== 'Active')
```
Comment (18alantom) [inline models/inventory/helpers.ts:237]: Braces!

File: models/inventory/helpers.ts [hunk 22]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+};
+
+const validateSNMaterialIssue = async (
+  doc: Doc,
+  name: string,
+  item: string,
+  serialNo: string,
+  status: string
+) => {
+  if (doc.isCancelled) return;
```
Comment (18alantom) [inline models/inventory/helpers.ts:232]: Braces!

File: models/inventory/helpers.ts [hunk 23]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+
+const validateSNShipment = async (doc: Doc, serialNo: string) => {
+  const { status } = await doc.fyo.db.get(
+    ModelNameEnum.SerialNo,
+    serialNo,
+    'status'
+  );
+
+  if (status !== 'Active')
+    throw new ValidationError(t`Serial No ${serialNo} status is not Active`);
```
Comment (18alantom) [inline models/inventory/helpers.ts:258]: Consistency! why is it Serial No here and Serial Number everywhere else?

Comment (18alantom) [reply models/inventory/helpers.ts:258]: Message should be "Serial Number ... is not Active"

Comment (18alantom) [reply models/inventory/helpers.ts:258]: Also absence of `"."` in some of the messages.

File: models/inventory/helpers.ts [hunk 24]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+      status: 'Inactive',
+    };
+    (
+      await doc.fyo.doc
+        .getNewDoc(ModelNameEnum.SerialNo, values as DocValueMap)
+        .sync()
+    ).submit();
+  }
+
+  if (status && status !== 'Inactive') {
```
Comment (18alantom) [inline models/inventory/helpers.ts:220]: Redundant check, if status is not defined then it won't be `"Inactive"`

File: models/inventory/helpers.ts [hunk 25]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+      name: serialNo,
+      item: doc.item,
+      party: doc.parentdoc?.party as string,
+      status: 'Inactive',
+    };
+    (
+      await doc.fyo.doc
+        .getNewDoc(ModelNameEnum.SerialNo, values as DocValueMap)
+        .sync()
+    ).submit();
```
Comment (18alantom) [inline models/inventory/helpers.ts:217]: Again, side effects, either bad function naming or it should be two functions.

File: models/inventory/helpers.ts [hunk 26]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+  }
+};
+
+const validateSNPurchaseReceipt = async (
+  doc: Doc,
+  name: string,
+  serialNo: string,
+  status: string
+) => {
+  if (name === undefined) {
```
Comment (18alantom) [inline models/inventory/helpers.ts:206]: Why are you even checking if name will be undefined here? The calling function already has a type restraint that `name` has to be a string!

This is bad!

File: models/inventory/helpers.ts [hunk 27]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+  if (status && status !== 'Inactive') {
+    throw new ValidationError(t`SerialNo ${serialNo} status is not Inactive`);
+  }
+};
+
+const validateSNPurchaseReceipt = async (
+  doc: Doc,
+  name: string,
+  serialNo: string,
+  status: string
```
Comment (18alantom) [inline models/inventory/helpers.ts:204]: Why isn't status typed as `SerialNumberStatus` you have defined it somewhere

File: models/inventory/helpers.ts [hunk 28]
```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
...
+    };
+    (
+      await doc.fyo.doc
+        .getNewDoc(ModelNameEnum.SerialNo, values as DocValueMap)
+        .sync()
+    ).submit();
+  }
+
+  if (status && status !== 'Inactive') {
+    throw new ValidationError(t`SerialNo ${serialNo} status is not Inactive`);
```
Comment (18alantom) [inline models/inventory/helpers.ts:221]: Why isn't there a space between Serial and No?

**Comment by 18alantom** [comment]
Doesn't LGTM, merging cause will be fixing issues in the next PR.
